test/e2e/utils/tls.go

```diff
@@ -0,0 +1,247 @@
+		cert:       cert,
+		certBytes:  certBytes,
+	}
+}
+
+func cert(name string, ca *certHolder) *certHolder {
+	// create our private and public key
+	privateKey, err := rsa.GenerateKey(rand.Reader, keyBits)
+	ExpectNoError(err)
+
+	// set up our server certificate
+	server := &x509.Certificate{
+		SerialNumber: big.NewInt(serialNumber),
+		Subject: pkix.Name{
+			CommonName:         name,
+			Organization:       []string{"JBoss"},
+			OrganizationalUnit: []string{"Infinispan"},
+			Locality:           []string{"Red Hat"},
+		},
+		Issuer:             ca.cert.Subject,
+		NotBefore:          time.Now(),
+		NotAfter:           time.Now().AddDate(10, 0, 0),
+		PublicKeyAlgorithm: x509.RSA,
+		SignatureAlgorithm: x509.SHA256WithRSA,
+	}
+	serialNumber++
+
+	certBytes, err := x509.CreateCertificate(rand.Reader, server, ca.cert, &privateKey.PublicKey, ca.privateKey)
+	ExpectNoError(err)
+
+	cert, err := x509.ParseCertificate(certBytes)
+	ExpectNoError(err)
+
+	return &certHolder{
+		privateKey: privateKey,
+		cert:       cert,
+		certBytes:  certBytes,
+	}
+}
+
+func createKeystore(ca, server *certHolder) []byte {
+	// It's not possible to use the p12 library as we get the following error
+	// TLS handshake failed: javax.net.ssl.SSLException: error:1417A0C1:SSL routines:tls_post_process_client_hello:no shared cipher
+	// keystore, err := p12.Encode(rand.Reader, server.privateKey, server.cert, []*x509.Certificate{ca.cert}, KeystorePassword)
+	var fileMode os.FileMode = 0777
+	ExpectNoError(os.MkdirAll(tmpDir, fileMode))
+	defer os.RemoveAll(tmpDir)
+
+	privKeyFile := tmpFile("server_key.pem")
+	certFile := tmpFile("server_cert.pem")
+	keystorefile := tmpFile("keystore.p12")
+
+	err := ioutil.WriteFile(privKeyFile, server.getPrivateKeyPEM(), fileMode)
+	ExpectNoError(err)
+
+	err = ioutil.WriteFile(certFile, append(server.getCertPEM(), ca.getCertPEM()...), fileMode)
+	ExpectNoError(err)
+
+	cmd := exec.Command("openssl", "pkcs12", "-export", "-in", certFile, "-inkey", privKeyFile,
+		"-name", server.cert.Subject.CommonName, "-out", keystorefile, "-password", "pass:"+KeystorePassword, "-noiter", "-nomaciter")
+	ExpectNoError(cmd.Run())
+
+	keystore, err := ioutil.ReadFile(keystorefile)
+	ExpectNoError(err)
+	return keystore
+}
+
+func createTruststore(ca, client *certHolder, authenticate bool) []byte {
+	var trustCerts []*x509.Certificate
+	// Only add the client certificate to the truststore if we require authentication
+	if authenticate {
+		trustCerts = []*x509.Certificate{ca.cert, client.cert}
+	} else {
+		trustCerts = []*x509.Certificate{ca.cert}
+	}
+	truststore, err := p12.EncodeTrustStore(rand.Reader, trustCerts, "secret")
+	ExpectNoError(err)
+	return truststore
+}
+
+func tmpFile(name string) string {
+	return fmt.Sprintf("%s/%s", tmpDir, name)
+}
+
+// Return the private key in PEM format
+func (c *certHolder) getPrivateKeyPEM() []byte {
+	privKeyPEM := new(bytes.Buffer)
+	err := pem.Encode(privKeyPEM, &pem.Block{
+		Type:  "RSA PRIVATE KEY",
```

Comment:
We can introduce a constant or use value from the `k8s.io/client-go/util/keyutil/key.go` package